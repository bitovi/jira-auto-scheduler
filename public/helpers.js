export function makeObjectMapByKey(issues, key) {
  // could replace with native methods
  // return Object.fromEntries(issues.map((issue) => [issue[key], issue]));
  var map = {};
  issues.forEach((issue) => {
    map[issue[key]] = issue;
  });
  return map;
}

export function makeFilterByPropertyNotEqualToOneOfValues(property, values) {
  return function (issue) {
    return !values.includes(issue[property]);
  };
}

export function groupByKey(issues, key) {
  var map = {};
  issues.forEach((issue) => {
    var value = typeof key === "function" ? key(issue) : issue[key];

    if (!map[value]) {
      map[value] = [];
    }
    map[value].push(issue);
  });
  return map;
}

export function splitByHavingPropertyValue(issues, splitProperty) {
  return partition(issues, function (issue) {
    return issue[splitProperty];
  });
}

export function partition(array, filter, thisArg) {
  var falsy = [];
  var truthy = [];
  array.forEach((issue, i) => {
    if (filter.call(thisArg, issue, i)) {
      truthy.push(issue);
    } else {
      falsy.push(issue);
    }
  });
  return { falsy, truthy };
}

export function stringToArray(value) {
  if (typeof value === "string") {
    return [value];
  } else {
    return value;
  }
}
