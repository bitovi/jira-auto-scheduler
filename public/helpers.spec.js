import { describe, expect } from "@jest/globals";
import { groupByKey, makeObjectMapByKey } from "./helpers";

describe("helpers methods", () => {
  describe("makeObjectMapByKey", () => {
    it("should convert array to dictionary", () => {
      let issues = [
        { key: "a", value: { name: "bob" } },
        { key: "b", value: { name: "ken" } },
        { key: "c", value: { name: "stu" } },
        { key: "z", value: { name: "mak" } },
        { key: "y", value: { name: "pat" } },
        { key: "x", value: { name: "cat" } },
      ];
      let key = "key";
      const actual = makeObjectMapByKey(issues, key);

      const match = Object.fromEntries(
        issues.map((issue) => [issue[key], issue])
      );
      expect(actual).toMatchSnapshot();
      expect(match).toMatchSnapshot();
    });
  });
  describe("groupByKey", () => {
    it("should convert array to dictionary of arrays", () => {
      let issues = [
        { key: "a", value: { name: "bob" } },
        { key: "a", value: { name: "ken" } },
        { key: "b", value: { name: "stu" } },
        { key: "b", value: { name: "mak" } },
        { key: "c", value: { name: "pat" } },
        { key: "c", value: { name: "cat" } },
      ];
      let key = "key";
      const actual = groupByKey(issues, key);

      expect(actual).toMatchSnapshot();
    });
  });
});
