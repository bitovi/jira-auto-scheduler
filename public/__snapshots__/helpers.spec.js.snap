// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers methods groupByKey should convert array to dictionary of arrays 1`] = `
{
  "a": [
    {
      "key": "a",
      "value": {
        "name": "bob",
      },
    },
    {
      "key": "a",
      "value": {
        "name": "ken",
      },
    },
  ],
  "b": [
    {
      "key": "b",
      "value": {
        "name": "stu",
      },
    },
    {
      "key": "b",
      "value": {
        "name": "mak",
      },
    },
  ],
  "c": [
    {
      "key": "c",
      "value": {
        "name": "pat",
      },
    },
    {
      "key": "c",
      "value": {
        "name": "cat",
      },
    },
  ],
}
`;

exports[`helpers methods makeObjectMapByKey should convert array to dictionary 1`] = `
{
  "a": {
    "key": "a",
    "value": {
      "name": "bob",
    },
  },
  "b": {
    "key": "b",
    "value": {
      "name": "ken",
    },
  },
  "c": {
    "key": "c",
    "value": {
      "name": "stu",
    },
  },
  "x": {
    "key": "x",
    "value": {
      "name": "cat",
    },
  },
  "y": {
    "key": "y",
    "value": {
      "name": "pat",
    },
  },
  "z": {
    "key": "z",
    "value": {
      "name": "mak",
    },
  },
}
`;

exports[`helpers methods makeObjectMapByKey should convert array to dictionary 2`] = `
{
  "a": {
    "key": "a",
    "value": {
      "name": "bob",
    },
  },
  "b": {
    "key": "b",
    "value": {
      "name": "ken",
    },
  },
  "c": {
    "key": "c",
    "value": {
      "name": "stu",
    },
  },
  "x": {
    "key": "x",
    "value": {
      "name": "cat",
    },
  },
  "y": {
    "key": "y",
    "value": {
      "name": "pat",
    },
  },
  "z": {
    "key": "z",
    "value": {
      "name": "mak",
    },
  },
}
`;
